### 5.2 그리드 서치
- 모델 매개변수 튜닝을 통한 일반화 성능 개선 
- 가장 널리 사용되는 방법은 Grid Search (그리드 탐색)
  - 관심있는 매개변수들을 대상으로 모든 조합을 시도함.

#### 5.2.1 간단한 그리드 서치
- SVC 모델에서 가장 중요한 매개변수는 gamma, C
- 그리드 탐색 범위 설정 예
  - gamma: [0.001, 0.01, 0.1, 1, 10, 100]
  - C: [0.001, 0.01, 0.1, 1, 10, 100]
  - 총 6x6=36개의 조합에 대하여 반복적으로 새로운 모델 생성 및 평가
  - 가장 좋은 성능을 보여주는 gamma와 C의 조합을 찾음

<table>
    <tr>
        <td>&nbsp;</td>
        <td>C=0.001</td>
        <td>C=0.01</td>
        <td>C=0.1</td>
        <td>C=1</td>
        <td>C=10</td>
        <td>C=100</td>
    </tr>
    <tr>
        <td>gamma=0.001</td>
        <td>SVC(C=0.001, gamma=0.001)</td>
        <td>SVC(C=0.01, gamma=0.001)</td>
        <td>SVC(C=0.1, gamma=0.001)</td>
        <td>SVC(C=1, gamma=0.001)</td>
        <td>SVC(C=10, gamma=0.001)</td>
        <td>SVC(C=100, gamma=0.001)</td>
    </tr>
    <tr>
        <td>gamma=0.01</td>
        <td>SVC(C=0.001, gamma=0.01)</td>
        <td>SVC(C=0.01, gamma=0.01)</td>
        <td>SVC(C=0.1, gamma=0.01)</td>
        <td>SVC(C=1, gamma=0.01)</td>
        <td>SVC(C=10, gamma=0.01)</td>
        <td>SVC(C=100, gamma=0.01)</td>
    </tr>
    <tr>
        <td>gamma=0.1</td>
        <td>SVC(C=0.001, gamma=0.1)</td>
        <td>SVC(C=0.01, gamma=0.1)</td>
        <td>SVC(C=0.1, gamma=0.1)</td>
        <td>SVC(C=1, gamma=0.1)</td>
        <td>SVC(C=10, gamma=0.1)</td>
        <td>SVC(C=100, gamma=0.1)</td>
    </tr>
    <tr>
        <td>gamma=1</td>
        <td>SVC(C=0.001, gamma=1)</td>
        <td>SVC(C=0.01, gamma=1)</td>
        <td>SVC(C=0.1, gamma=1)</td>
        <td>SVC(C=1, gamma=1)</td>
        <td>SVC(C=10, gamma=1)</td>
        <td>SVC(C=100, gamma=1)</td>
    </tr>
    <tr>
        <td>gamma=10</td>
        <td>SVC(C=0.001, gamma=10)</td>
        <td>SVC(C=0.01, gamma=10)</td>
        <td>SVC(C=0.1, gamma=10)</td>
        <td>SVC(C=1, gamma=10)</td>
        <td>SVC(C=10, gamma=10)</td>
        <td>SVC(C=100, gamma=10)</td>
    </tr>
    <tr>
        <td>gamma=100</td>
        <td>SVC(C=0.001, gamma=100)</td>
        <td>SVC(C=0.01, gamma=100)</td>
        <td>SVC(C=0.1, gamma=100)</td>
        <td>SVC(C=1, gamma=100)</td>
        <td>SVC(C=10, gamma=100)</td>
        <td>SVC(C=100, gamma=100)</td>
    </tr>    
</table>

In [1]:
%matplotlib inline
import sys 
sys.path.append('..')
from preamble import *

from sklearn.datasets import make_blobs
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# naive grid search implementation
from sklearn.svm import SVC
from sklearn.datasets import load_iris
iris = load_iris()

X_train, X_test, y_train, y_test = train_test_split(iris.data, iris.target, random_state=0)
print("Size of training set: {}   size of test set: {}".format(X_train.shape[0], X_test.shape[0]))

best_score = 0

for gamma in [0.001, 0.01, 0.1, 1, 10, 100]:
    for C in [0.001, 0.01, 0.1, 1, 10, 100]:
        # for each combination of parameters, train an SVC
        svm = SVC(gamma=gamma, C=C)
        svm.fit(X_train, y_train)
        # evaluate the SVC on the test set
        score = svm.score(X_test, y_test)
        # if we got a better score, store the score and parameters
        if score > best_score:
            best_score = score
            best_parameters = {'C': C, 'gamma': gamma}

print("Best score: {:.2f}".format(best_score))
print("Best parameters: {}".format(best_parameters))

Size of training set: 112   size of test set: 38
Best score: 0.97
Best parameters: {'C': 100, 'gamma': 0.001}


- 위 0.97의 정확도는 전혀 새로운 데이터에 대한 성능으로 이어지지 않을 수 있다.
- 즉, 위 예제에서 사용한 테스트 데이터는 모델 구성시에 사용을 해버렸기 때문에 이 모델이 얼마나 좋은지 평가하는 데 더 이상 사용할 수 없다.